File: materialHardship/data_prepare.py

# coding: utf-8

# # Prepare data (just for materialHardship)

# In[2]:

import numpy as np
import pandas as pd
# import xgboost as xgb
# from sklearn.model_selection import train_test_split
import matplotlib.pyplot as plt
import seaborn as sns


# In[3]:

np.random.seed(1234)
print('reading in csv files')
# File read in
# See documentation for more descriptions
# http://www.fragilefamilieschallenge.org/apply/
background = pd.read_csv('background.csv', low_memory=False)
print('read background.csv')
train = pd.read_csv('train.csv', low_memory=False)
print('read train.csv')
prediction = pd.read_csv('prediction.csv', low_memory=False)
print('read prediction.csv')
print("Files read")


# Background is super WIDE. We have only 4242 IDs, but almost 13k features!

# In[4]:

background.shape


# OK...so they want predictions for...everything. So we are going to make a **training** set using the rows of background that have non-NA values for materialHardship.
# 
# Then we need to make predictions for every challengeID that is a) not train or b) is in train by has NA for materialHardship 

# In[5]:

all_y_train_materialHardship = train[['challengeID', 'materialHardship']]
# non NA y_train ds and data
non_na_y_train_materialHardship = all_y_train_materialHardship.dropna()
non_na_y_train_materialHardship.head()


# In[6]:

# y_train_ids that are na
na_y_train_materialHardship = all_y_train_materialHardship.loc[~all_y_train_materialHardship['challengeID'].isin(non_na_y_train_materialHardship['challengeID'])]
na_y_train_materialHardship.head()


# Our training set will have all ids that we have non-NA material hardship data for. 

# In[7]:

df_train = background.loc[background['challengeID'].isin(non_na_y_train_materialHardship['challengeID'])]
df_train.head()


# The prediction set contains all rows in 'background.csv' that are not in df_train.

# In[9]:

df_prediction = background.loc[~background['challengeID'].isin(df_train['challengeID'])]
df_prediction.sort_values(by='challengeID').head(10)


# SO, we have the data that we will train on:
#     
#     - Training features: 'df_train' 
#     - Training labels: 'non_na_y_train_materialHardship'
# 
# We'll use this model to predict output (materialHardship) values for:
# 
#     - Prediction features: 'df_prediction'

# ### df_train

# In[11]:

df_train.head()


# ### non_na_y_train_materialHardship

# In[12]:

non_na_y_train_materialHardship.head()


# ### df_prediction

# In[13]:

df_prediction.head()


# We should have that the sum of rows in df_train and df_prediction sum to all of the rows in background.csv. Let's ensure that.

# In[15]:

np.sum(df_train.shape[0]+df_prediction.shape[0])==background.shape[0]


# OK. We are off to the races and can reference these pandas dataframes in future data preparation and training notebooks!

# In[ ]:

File: materialHardship/data-prepare.R
## Prepare data just for materialHardship

library(data.table)
library(tidyverse)

#Not using randomness here, but whatever
set.seed(1234)

#This assumes you're in the working directory with the files.
print("Reading in CSV files. Why doesn't Will capitalize words?")

background <- fread("background.csv")
prediction <- fread("prediction.csv")
train      <- fread("train.csv")

print("Files read. Why doesn't Will punctuate sentences?")

# Background is super WIDE. 
# We have only 4242 IDs, but almost 13k features!
# Yikes!!

dim(background); dim(prediction); dim(train)

#OK...so they want predictions for...everything.
# So we are going to make a **training** set,
# using the rows of background that
# have non-NA values for materialHardship. Fine.
# 

train.no.na.mh <- train[(!is.na(materialHardship))]

# Then we need to make predictions for every challengeID that is
# (a) not in this training set, or
# (b) is in the training set, but has NA for materialHardship 

prediction.ids <- unique(c( train[is.na(materialHardship),challengeID],
                     background[,setdiff(challengeID,
                                         train.no.na.mh[,challengeID])]))

stopifnot( length(prediction.ids) + nrow(train.no.na.mh) == nrow(background) )

#OK. Now pull these ideas from the background set

#keying and joining are important data.table concepts.
#This basically means "tell background to organize itself so that
# challengeID is the identifying variable for a row,
setkey(background,challengeID)
# then look up all the challengeIDs that match prediction.ids
# and pull those records.
background.subset.for.prediction <- background[J(prediction.ids)]
#See? Remark that there as many rows as length(prediction.ids)
dim(background.subset.for.prediction)
length(prediction.ids)

#Now let's do the same with the IDs we *are* going to train on
#Here, 
train.ids <- train.no.na.mh[,challengeID]
background.subset.for.training <- background[J(train.ids)]

#So now we have two sets of data, one for training the model,
#and one for out-of-sample predictions. 
#Notice that the row dimensions add up:

dim(background.subset.for.training)
dim(background.subset.for.prediction)

stopifnot(
    nrow(background.subset.for.training) + 
    nrow(background.subset.for.prediction)  ==
    nrow(background) 
)

#Now write these suckers out for easy use later.

fwrite(background.subset.for.training,file="background-train_only.csv")
fwrite(background.subset.for.prediction,file="background-predict_only.csv")
